alerts/send.go

```diff
@@ -0,0 +1,94 @@
+package alerts
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/sideshow/apns2"
+	"github.com/sideshow/apns2/payload"
+	"github.com/sideshow/apns2/token"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/errors"
+)
+
+// Pusher abstracts the underlying push notification mechanism.
+type Pusher interface {
+	// Push a notification to a device.
+	//
+	// deviceToken should uniquely identify a device, while Notification
+	// abstracts the provider-specific notification payload.
+	Push(ctx context.Context, deviceToken devicetokens.DeviceToken, notification *Notification) error
+}
+
+// Notification models a provider-independent push notification.
+type Notification struct {
+	Title   string
+	Message string
+}
+
+// APNSPusher implements push notifications via Apple APNs.
+type APNSPusher struct {
+	serviceToken *token.Token
+	bundleID     string
+}
+
+// NewAPNSPusher creates a Pusher for sending device notifications via Apple's
+// APNs.
+//
+// The signingKey is the raw token signing key received from Apple (.p8 file
+// containing PEM-encoded private key), along with its respective team id, key
+// id, and application bundle id.
+//
+// https://developer.apple.com/documentation/usernotifications/sending-notification-requests-to-apns
+func NewAPNSPusher(signingKey []byte, keyID, teamID, bundleID string) (*APNSPusher, error) {
+	authKey, err := token.AuthKeyFromBytes(signingKey)
+	if err != nil {
+		return nil, err
+	}
+	token := &token.Token{
+		AuthKey: authKey,
+		KeyID:   keyID,
+		TeamID:  teamID,
+	}
+	return &APNSPusher{
+		bundleID:     bundleID,
+		serviceToken: token,
+	}, nil
+}
+
+func (p *APNSPusher) Push(ctx context.Context, deviceToken devicetokens.DeviceToken, note *Notification) error {
+	if deviceToken.Apple == nil {
+		return errors.New("Unable to push notification: APNSPusher can only use Apple device tokens but the Apple token is nil")
+	}
+
+	// TODO: look at the clientmanager package in apns2
+	client := apns2.NewTokenClient(p.serviceToken)
```

Comment:
We'll want to create this client once and keep it around. From the APNS2 documentation:

> As per the Apple APNs Provider API, you should keep a handle on this client so that you can keep your connections with APNs open across multiple notifications; don’t repeatedly open and close connections. APNs treats rapid connection and disconnection as a denial-of-service attack.

I think this implies a higher-level "manager" object that is created at the beginning of time for the service and maintains the client.

Reply:
And, to support multiple simultaneous outgoing PNs, you'll want to look into the APNS2 client manager (as your TODO comment states above).

Reply:
Yep, a good call out. I'm aware. This was just the shortest/easiest path to something that worked. :)

I'll be sure to revise this before we're ready to get live.

Reply:
> I think this implies a higher-level "manager" object that is created at the beginning of time for the service and maintains the client.

The manager just tracks a list of clients, and can remove old ones (it's not clear why a client being old is a problem, they auto-reconnect). More importantly for us, it only works with certificate clients, while we're using token clients. In other words, the client manager won't work for our use case.

I don't think we need more than a single client for now, but if we decide we need multiple later, we can use the sync.Pool or something along those lines.